File: chapter2/app/app/src/main/java/net/numa08/kotlinbook/chapter2/repositories/impl/ApplicationInformationRepositoryImpl.java
package net.numa08.kotlinbook.chapter2.repositories.impl;

import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.graphics.Palette;

import net.numa08.kotlinbook.chapter2.models.ApplicationInformation;
import net.numa08.kotlinbook.chapter2.repositories.ApplicationInformationRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class ApplicationInformationRepositoryImpl implements ApplicationInformationRepository {

    private final PackageManager packageManager;

    public ApplicationInformationRepositoryImpl(PackageManager packageManager) {
        this.packageManager = packageManager;
    }

    @Override
    public void findAllApplications(final FindAllApplicationsCallback callback) {
        final Handler callbackHandler = new Handler(Looper.myLooper());
        new Thread(new Runnable() {
            @Override
            public void run() {
                final List<ApplicationInformation> applications = findAllApplications();
                callbackHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        callback.onFindAllApplications(applications);
                    }
                });
            }
        }).start();

    }

    @Override
    public void findApplicationByPackageName(final String packageName, final FindApplicationCallback callback) {
        final Handler callbackHandler = new Handler(Looper.myLooper());
        new Thread(new Runnable() {
            @Override
            public void run() {
                final ApplicationInformation applicationInformation = findApplicationByPackageName(packageName);
                callbackHandler.post(new Runnable() {
                    @Override
                    public void run() {
                        callback.onFindApplication(applicationInformation);
                    }
                });
            }
        }).start();
    }

    @Nullable
    private ApplicationInformation findApplicationByPackageName(String packgeName) {
        try {
            final ApplicationInfo applicationInfo = packageManager.getApplicationInfo(packgeName, 0);
            return convertApplicationInfo(applicationInfo);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
            return null;
        }
    }

    private List<ApplicationInformation> findAllApplications() {
        final ExecutorService exec = Executors.newFixedThreadPool(5);
        final List<Future<ApplicationInformation>> tasks = new ArrayList<>();
        final List<ApplicationInformation> apps = new ArrayList<>();
        for (final ApplicationInfo appInfo :
                packageManager.getInstalledApplications(0)) {
            if ((appInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 0) {
                continue;
            }
            final Future<ApplicationInformation> task = exec.submit(new Callable<ApplicationInformation>() {

                @Override
                public ApplicationInformation call() throws Exception {
                    return convertApplicationInfo(appInfo);
                }
            });
            tasks.add(task);
        }

        for (Future<ApplicationInformation> task :
                tasks) {
            //noinspection TryWithIdenticalCatches
            try {
                apps.add(task.get());
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }
        }
        exec.shutdown();
        return apps;
    }

    @NonNull
    private ApplicationInformation convertApplicationInfo(ApplicationInfo appInfo) {
        final CharSequence label = appInfo.loadLabel(packageManager);
        final Drawable icon = appInfo.loadIcon(packageManager);
        final CharSequence description = appInfo.loadDescription(packageManager);
        final int lightVibrantRGB;
        final int vibrantRGB;
        final int bodyTextColor;
        final int titleTextColor;
        if (icon instanceof BitmapDrawable) {
            final Bitmap bitmap = ((BitmapDrawable) icon).getBitmap();
            final Palette palette = new Palette.Builder(bitmap).generate();
            final Palette.Swatch lightVibrant = palette.getLightVibrantSwatch();
            if (lightVibrant != null) {
                lightVibrantRGB = lightVibrant.getRgb();
                bodyTextColor = lightVibrant.getBodyTextColor();
                titleTextColor = lightVibrant.getTitleTextColor();
            } else {
                lightVibrantRGB = bodyTextColor = titleTextColor = 0;
            }
            final Palette.Swatch vibrantSwatch = palette.getVibrantSwatch();
            if (vibrantSwatch != null) {
                vibrantRGB = vibrantSwatch.getRgb();
            } else {
                vibrantRGB = 0;
            }
        } else {
            lightVibrantRGB = vibrantRGB = bodyTextColor = titleTextColor = 0;
        }
        return new ApplicationInformation(
                label,
                icon,
                description,
                lightVibrantRGB,
                vibrantRGB,
                bodyTextColor,
                titleTextColor,
                appInfo.packageName,
                appInfo);
    }
}

File: chapter2/app/app/src/main/java/net/numa08/kotlinbook/chapter2/repositories/impl/ApplicationInformationRepositoryImpl.kt
package net.numa08.kotlinbook.chapter2.repositories.impl

import android.annotation.SuppressLint
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.drawable.BitmapDrawable
import android.os.Handler
import android.os.Looper
import android.support.annotation.VisibleForTesting
import android.support.v7.graphics.Palette
import net.numa08.kotlinbook.chapter2.models.ApplicationInformation
import net.numa08.kotlinbook.chapter2.repositories.ApplicationInformationRepository
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors

class ApplicationInformationRepositoryImpl(private val packageManager: PackageManager) : ApplicationInformationRepository {

    @SuppressLint("VisibleForTests")
    override fun findAllApplications(cb: ((List<ApplicationInformation>) -> Unit)) {
        val callbackHandler = Handler(Looper.myLooper())
        Thread(Runnable {
            val applications = findAllApplications()
            callbackHandler.post { cb(applications) }
        }).start()

    }

    @SuppressLint("VisibleForTests")
    override fun findApplicationByPackageName(packageName: String, cb: (ApplicationInformation?) -> Unit) {
        val callbackHandler = Handler(Looper.myLooper())
        Thread(Runnable {
            val applicationInformation = findApplicationByPackageName(packageName)
            callbackHandler.post { cb(applicationInformation) }
        }).start()
    }

    @VisibleForTesting
    fun findApplicationByPackageName(packageName: String): ApplicationInformation? {
        return try {
            val applicationInfo = packageManager.getApplicationInfo(packageName, 0)
            convertApplicationInfo(applicationInfo)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
            null
        }
    }

    @VisibleForTesting
    fun findAllApplications(): List<ApplicationInformation> {
        val exec = Executors.newFixedThreadPool(5)
        val apps =
                packageManager.getInstalledApplications(0)
                        .filter { it.flags and ApplicationInfo.FLAG_SYSTEM == 0 }
                        .map { exec.submit(Callable { convertApplicationInfo(it) }) }
                        .mapNotNull { task ->
                            try {
                                task.get()
                            } catch (e: InterruptedException) {
                                e.printStackTrace()
                                null
                            } catch (e: ExecutionException) {
                                e.printStackTrace()
                                null
                            }
                        }
        exec.shutdown()
        return apps
    }

    @VisibleForTesting
    fun convertApplicationInfo(appInfo: ApplicationInfo): ApplicationInformation {
        val label = appInfo.loadLabel(packageManager)
        val icon = appInfo.loadIcon(packageManager)
        val description = appInfo.loadDescription(packageManager)
        val lightVibrantRGB: Int?
        val vibrantRGB: Int?
        val bodyTextColor: Int?
        val titleTextColor: Int?
        if (icon is BitmapDrawable) {
            val bitmap = icon.bitmap
            val palette = Palette.Builder(bitmap).generate()
            val lightVibrant = palette.lightVibrantSwatch
            lightVibrantRGB = lightVibrant?.rgb
            bodyTextColor = lightVibrant?.bodyTextColor
            titleTextColor = lightVibrant?.titleTextColor
            val vibrantSwatch = palette.vibrantSwatch
            vibrantRGB = vibrantSwatch?.rgb
        } else {
            titleTextColor = null
            bodyTextColor = titleTextColor
            vibrantRGB = bodyTextColor
            lightVibrantRGB = vibrantRGB
        }
        return ApplicationInformation(
                label,
                icon,
                description,
                lightVibrantRGB,
                vibrantRGB,
                bodyTextColor,
                titleTextColor,
                appInfo.packageName,
                appInfo)
    }
}
